cava/CoqStyleGuide.md

```diff
@@ -0,0 +1,242 @@
+the [example](#Example) at the bottom of this file.
+
+## Code organization
+
+### Legal banner
+
+- Files should begin with a copyright/license banner, as shown in the example
+  above.
+
+### Import statements
+
+- `Require Import` statements should all go at the top of the file, followed by
+  file-wide `Import` statements.
+  * `Import`s often contain notations or typeclass instances that might
+     override notations or instances from another library, so it's nice to
+     highlight them separately.
+- One `Require Import` statement per line; it's easier to scan that way.
+- `Require Import` statements should use "fully-qualified" names (e.g. `Require Import Coq.ZArith.ZArith` instead of `Require Import ZArith`).
+  * Use the `Locate` command to find the fully-qualified name!
+- `Require Import`s should go in the following order:
+   1. Standard library dependencies (start with `Coq.`)
+   2. External dependencies (anything outside the current project)
+   3. Same-project dependencies
+- `Require Import`s with the same root library (the name before the first `.`)
+should be grouped together. Within each root-library group, they should be in
+alphabetical order (so `Coq.Lists.List` before `Coq.ZArith.ZArith`).
+
+### Notations and scopes
+
+- Any file-wide `Local Open Scope`s should come immediately after the `Import`s
+  (see example).
+  * Always use `Local Open Scope`; just `Open Scope` will sneakily
+open the scope for those who import your file.
+- Put notations in their own separate modules or files, so that those who import your file can choose whether or not they want the notations.
+  * Conflicting notations can cause a lot of headache, so it comes in very handy to leave this flexibility!
+
+## Formatting
+
+### Line length
+
+- Maximum line length 80 characters.
+  * Many Coq IDE setups divide the screen in half vertically and use only half
+    to display source code, so more than 80 characters can be genuinely hard to
+    read on a laptop.
+
+### Whitespace and indentation
+
+- No trailing whitespace.
+- Spaces, not tabs.
+- Files should end with a newline.
+  * Many editors do this automatically on save.
+- Default indentation is 2 spaces.
+  * Keeping this small prevents complex proofs from being indented ridiculously
+    far, and matches IDE defaults.
+- Use 2-space indents if inserting a line break immediately after:
+  * `Proof.`
+  * `fun <...> =>`
+  * `forall <...>,`
+  * `exists <....>,`
+- The style for indenting arguments in function application depends on where
+  you make a line break.  If you make the line break immediately after the
+function name, use a 2-space indent.  However, if you make it after one or more
+arguments, align the next line with the first argument:
+  ```coq
+  (Z.pow
+     1 2)
+  (Z.pow 1 2 3
+         4 5 6)
+  ```
+- `Inductive` cases should not be indented. Example:
+  ```coq
+  Inductive Foo : Type :=
+  | FooA : Foo
+  | FooB : Foo
+  .
```

Comment:
Should the `.` always be on a line by itself at the end? Likewise for a type with just one constructor?

Reply:
I always put it on a line by itself at the end, because that makes it easy to add/remove constructors. I was debating putting it in, but it seemed maybe too picky for a style guide. What do you think?

Reply:
I usually surround colons with spaces, but I don't feel strongly about it either, and am happy to leave it up to individual preference. Maybe for these things we want to leave unspecified, to minimize ambiguity, we should add a point to the guide saying explicitly that either way is okay?